pandas/core/arrays/boolean.py

```diff
@@ -0,0 +1,605 @@
+
+    mask_values = None
+    if isinstance(values, np.ndarray) and values.dtype == np.bool_:
+        if copy:
+            values = values.copy()
+    else:
+        values_object = np.asarray(values, dtype=object)
+        mask_values = isna(values)
+        values = np.zeros(len(values), dtype=bool)
+        values[~mask_values] = values_object[~mask_values].astype(bool)
+
+    if isinstance(mask, np.ndarray) and mask.dtype == np.bool_:
+        if copy:
+            mask = mask.copy()
+    else:
+        if mask is None:
+            if mask_values is None:
+                # TODO optimize this case?
+                mask = np.zeros(len(values), dtype=bool)
+            else:
+                mask = mask_values
+        else:
+            mask = np.asarray(mask, dtype=bool)
+            if mask_values is not None:
+                mask = mask | mask_values
+
+    if not values.ndim == 1:
+        raise TypeError("values must be a 1D list-like")
+    if not mask.ndim == 1:
+        raise TypeError("mask must be a 1D list-like")
+
+    return values, mask
+
+
+class BooleanArray(ExtensionArray, ExtensionOpsMixin):
+    def __init__(self, values, mask, copy=False):
+        if not (isinstance(values, np.ndarray) and values.dtype == np.bool_):
+            raise TypeError(
+                "values should be boolean numpy array. Use "
+                "the 'array' function instead"
+            )
+        if not (isinstance(mask, np.ndarray) and mask.dtype == np.bool_):
+            raise TypeError(
+                "mask should be boolean numpy array. Use "
+                "the 'array' function instead"
+            )
+
+        if copy:
+            values = values.copy()
+            mask = mask.copy()
+
+        self._data = values
+        self._mask = mask
+        self._dtype = BooleanDtype()
+
+    @property
+    def dtype(self):
+        return self._dtype
+
+    @classmethod
+    def _from_sequence(cls, scalars, dtype=None, copy=False):
+        if dtype:
+            assert dtype == "boolean"
+        values, mask = coerce_to_array(scalars, copy=copy)
+        return BooleanArray(values, mask)
+
+    @classmethod
+    def _from_factorized(cls, values, original):
+        return cls._from_sequence(values, dtype=original.dtype)
+
+    def _formatter(self, boxed=False):
+        def fmt(x):
+            if isna(x):
+                return "NaN"
+            return str(x)
+
+        return fmt
+
+    def __getitem__(self, item):
+        if is_integer(item):
+            if self._mask[item]:
+                return self.dtype.na_value
+            return self._data[item]
+        return type(self)(self._data[item], self._mask[item])
+
+    def _coerce_to_ndarray(self):
+        """
+        coerce to an ndarary of object dtype
+        """
+        if self.isna().any():
+            data = self._data.astype(object)
+            data[self._mask] = self._na_value
+            return data
+        else:
+            return self._data
```

Comment:
I adapted this compared to IntegerArray to return bool dtype if there are no NaNs (this helps to enable boolean indexing automatically for this case). But we should probably make a consistent choice.

Reply:
Yeah, I don't think we want to return object-dtype here if we can avoid it...

Reply:
On the other hand, a predictable dtype (not depending on the values in the array) is also nice ...

Reply:
Yeah, as I thought more about it the dtype stability was winning me over. Not sure what's best.

Reply:
Switched back to always object dtype for now. There are methods (like astype(bool)) that force you to give bool dtype if possible. We should maybe think about adding a `to_numpy` to our arrays as well (the same for IntegerArray).